local/sdk-sriov/pkg/deviceplugin/server.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+const (
+	resourceNamePrefix = "networkservicemesh.io/"
+	kubeletNotifyDelay = 30 * time.Second
+	vfioDir            = "/dev/vfio"
+)
+
+// ServerConfig is a struct for configuring device plugin server
+type ServerConfig struct {
+	ResourceName  string
+	ResourceCount int
+	HostBaseDir   string
+	HostPathEnv   string
+}
+
+type devicePluginServer struct {
+	resourceName  string
+	resourceCount int
+	hostBaseDir   string
+	hostPathEnv   string
+	ctx           context.Context
+}
+
+// StartServer creates a new SR-IOV forwarder device plugin server and starts it
+func StartServer(ctx context.Context, config *ServerConfig, manager Manager) error {
+	logFmt := "StartServer: %v"
+
+	s := &devicePluginServer{
+		resourceName:  resourceNamePrefix + config.ResourceName,
+		resourceCount: config.ResourceCount,
+		hostBaseDir:   config.HostBaseDir,
+		hostPathEnv:   config.HostPathEnv,
+		ctx:           ctx,
+	}
+
+	logrus.Infof(logFmt, "starting server")
+	socket, err := manager.StartDeviceServer(s.ctx, s)
+	if err != nil {
+		logrus.Errorf(logFmt, "error starting server")
+		return err
+	}
+
+	logrus.Infof(logFmt, "registering server")
+	if err := s.register(manager, socket); err != nil {
+		logrus.Errorf(logFmt, "error registering server")
+		return err
+	}
+
+	if resetCh, err := manager.MonitorKubeletRestart(s.ctx); err == nil {
+		go func() {
+			logrus.Infof(logFmt, "start monitoring kubelet restart")
+			defer logrus.Infof(logFmt, "stop monitoring kubelet restart")
+			for {
+				select {
+				case <-s.ctx.Done():
+					return
+				case _, ok := <-resetCh:
+					if !ok {
+						return
+					}
+					logrus.Infof(logFmt, "re registering server")
+					if err := s.register(manager, socket); err != nil {
+						logrus.Errorf(logFmt, fmt.Sprint("error re registering server: ", err))
+						return
+					}
+				}
+			}
+		}()
+	} else {
+		logrus.Warnf(logFmt, "error monitoring kubelet restart")
```

Comment:
Do we need to log err here?

Reply:
This is not actually an error I suppose. We still can start and register server, but we lack monitoring for kubelet restart and so we can't re register server in such situation.
So it is not good, but we can work - log warn is the thing for it.

Reply:
I meant `err` from line 70.

Reply:
Yep, got it.